### PR TITLE
network: fix infinite loop in setting up wireguard

### LINKED ISSUE #11404: 23 wireguard peers hang systemd-networkd
**systemd version the issue has been seen with**
> 239-7ubuntu10.6

**Used distribution**
> Ubuntu 18.10

**Expected behaviour you didn't see**
> `systemctl restart systemd-networkd` restarts cleanly and applies network configuration

**Unexpected behaviour you saw**
> `systemctl restart systemd-networkd` hangs for a while and fails to apply network configuration. If the node is rebooted, it can't apply network configuration on boot as well.

**Steps to reproduce the problem**
1. launch VM on Linode with Ubuntu 18.04 or 18.10
1. add wireguard PPA: `apt-add-repository ppa:wireguard/wireguard`
1. install wireguard: `apt-get install wireguard`
1. download config file generator [`systemd-wg`](https://launchpadlibrarian.net/405947185/systemd-wg) that reproduces the issue
1. simulate my VPN concentrator configuration with `./systemd-wg 23`

For some reason, 23 is the magic number of peers: 22 peers, 24 peers and 1500 peers work okay, but 23 peers trigger the bug.

The bug also affects [systemd-237 from Ubuntu 18.04](https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1811149).

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #11404.

### CODE REVIEW

**Comment by yuwata** [comment]
@Mic92 Please take a look.

**Comment by Mic92** [comment]
Looks less scary then the old code.
I had look on the code and tested the build with some peers.

cc @zx2c4 in case you are interested.

**Comment by poettering** [comment]
looks ok to me, but i'd love to see the CI pass on this one before we merge it?

File: src/network/netdev/wireguard.c [hunk 1]
```diff
@@ -45,22 +45,137 @@ static WireguardPeer *wireguard_peer_new(Wireguard *w, unsigned section) {
...
+
+        /* This returns 1 on success, 0 on recoverable error, and negative errno on failure. */
+
+        r = sd_netlink_message_open_array(message, index);
+        if (r < 0)
+                return 0;
+
+        r = sd_netlink_message_append_u16(message, WGALLOWEDIP_A_FAMILY, mask->family);
+        if (r < 0)
+                goto cancel;
```
Comment (keszybz) [inline src/network/netdev/wireguard.c:63]: I don't understand the logic here. If this fails, the `for` loop in caller is aborted, so the error is silently ignored. Shouldn't this just error out?

Comment (Mic92) [reply src/network/netdev/wireguard.c:63]: The idea is that netlink's messages are limited in size. If an interface has many peers, addresses or ip masks then the configuration might not fit into one message and has to be split across different messages.

Comment (yuwata) [reply src/network/netdev/wireguard.c:63]: Yeah. I guess there was some bug in the cancellation logic, and it causes infinite loop with the magic number 23.

**Comment by yuwata** [comment]
bionic-amd64 failure is at `boot-and-services`
```
test_no_options (__main__.JournalTest) ... FAIL
```
and it seems not related to this PR. See https://salsa.debian.org/systemd-team/systemd/blob/master/debian/tests/boot-and-services#L176-183

**Pull request merged**